File: Programming_cheat_sheets/_scripting_languages/Ruby/data_structures/graph/graph.rb
# http://www.geeksforgeeks.org/graph-and-its-representations/

=begin 
    A Ruby Program to demonstrate adjacency list representation of graphs
    - vertex, vertices are the nodes
    - edges are the connections between the nodes

    - there are directed, and undirected graphs
    - acyclic and cyclic graphs
=end

# A class to represent an adjacency list node
class AdjListNode
    attr_accessor :dest, :next
end

# A class to represent an adjacency list
class AdjList
    attr_accessor :head  # pointer to head node of list
end

# A class to represent a graph. A graph is an array of adjacency lists.
# Size of array will be for the number of vertices in graph
class Graph
    attr_accessor :vertecesCount , :adjacency_lists  #:adjacency_lists
end

# A utility function to create a new adjacency list node
def new_adj_list_node(dest)
    new_node = AdjListNode.new
    new_node.dest = dest
    new_node.next = nil
    return new_node
end

# A utility function that creates a graph with n vertices
def createGraph(vertecesCount)
    graph = Graph.new
    graph.vertecesCount = vertecesCount

    # Create an array of adjacency lists.  Size of array will be the vertecesCount
    graph.adjacency_lists = []  # (struct AdjList *) malloc(vertecesCount * sizeof(struct AdjList));

    # Initialize each adjacency list as empty by making head as nil
    vertecesCount.times do |i|
        adj_list = AdjList.new
        adj_list.head = nil
        graph.adjacency_lists[i] = adj_list
    end
    return graph
end

# Adds an edge to an undirected graph
def addEdge(graph, src, dest)
    # Add an edge from src to dest.  A new node is added to the adjacency
    # list of src.  The node is added at the begining
    new_node = new_adj_list_node(dest)
    new_node.next = graph.adjacency_lists[src].head
    graph.adjacency_lists[src].head = new_node

    # Since graph is undirected, add an edge from dest to src also
    new_node = new_adj_list_node(src)
    new_node.next = graph.adjacency_lists[dest].head
    graph.adjacency_lists[dest].head = new_node
end

# A utility function to print the adjacenncy list representation of graph
def printGraph(graph)
    graph.adjacency_lists.each_with_index do |v, index|
        current_vertex = v.head
        print "\n Adjacency list of vertex #{index}\n head"
        while (current_vertex) do
            print " -> #{current_vertex.dest}"
            current_vertex = current_vertex.next
        end
        puts
    end
end

#-------------------------------------------------------------------------

# create the graph
@graph = createGraph(5)
addEdge(@graph, 0, 1)
addEdge(@graph, 0, 4)
addEdge(@graph, 1, 2)
addEdge(@graph, 1, 3)
addEdge(@graph, 1, 4)
addEdge(@graph, 2, 3)
addEdge(@graph, 3, 4)

# print the adjacency list representation of the above graph
printGraph(@graph)

=begin  

    output:

    Adjacency list of vertex 0
    head -> 4 -> 1

    Adjacency list of vertex 1
    head -> 4 -> 3 -> 2 -> 0

    Adjacency list of vertex 2
    head -> 3 -> 1

    Adjacency list of vertex 3
    head -> 4 -> 2 -> 1

    Adjacency list of vertex 4
    head -> 3 -> 1 -> 0

=end

File: This&that/_scripting_languages/Javascript/_data_structures/graphs/graph.js
/*
 * graph
 *
 * https://www.syncano.io/blog/data-structures-in-javascript/
 *
 * it can be represented as a edge list, 
 * or adjancecy matrix ?
 *
 * vertex is a node
 *
 * edge is a connection between nodes 
 *
 * directed and undirected graphs
 *
 * degree of a node: number of connection of the node
 *
 * adjencency list or adjacency matrix https://www.youtube.com/watch?v=WtfGRS1BsBI
 *
 * adj list - array of linkd lists
 * adj matrix - 2D array
 *
 * self loops
 *
 * DFS BFS(needs a Queue)
 *
 */
function AdjListNode(dest){
    this.dest = dest;
    this.next = null;
}

function AdjList(){
    this.head = null;
}

function Graph(n){
    this.nodeCOunt = 0;
    this.ajdLists = [];

    this.count = n;
    this.adjLists = [];

    for(var i = 0; i < n; i++){
        var adjList = new AdjList();
        adjList.head = null;
        this.ajdLists[i] = adjList;
    }

    this.newNode = function(dest){
        var new_node = new AdjListNode();
        new_node.dest = dest;
        new_node.next = null;
        return new_node;
    };

    this.addEdge = function(src, dest){
        var new_node = this.newNode(dest);
        new_node.next = this.ajdLists[src].head;
        this.ajdLists[src].head = new_node;

        // Since this is undirected, add an edge from dest to src also
        new_node = this.newNode(src);
        new_node.next = this.ajdLists[dest].head;
        this.ajdLists[dest].head = new_node;
    };

    this.printGraph = function(){
        this.ajdLists.forEach(function(v, index){
            var current_vertex = v.head;
            console.log("Adjacency list of vertex " + index + ":");
            var str = "HEAD(" + index + ")";
            while(current_vertex){
                str += " -> " + current_vertex.dest;
                current_vertex = current_vertex.next;
            }
            console.log(str + '\n');
        }); 
    };
}

var graph = new Graph(5);

graph.addEdge(0,1);
graph.addEdge(0,4);
graph.addEdge(1,2);
graph.addEdge(1,3);
graph.addEdge(1,4);
graph.addEdge(2,3);
graph.addEdge(3,4);

graph.printGraph();

/* output: 

Adjacency list of vertex 0:
HEAD(0) -> 4 -> 1

Adjacency list of vertex 1:
HEAD(1) -> 4 -> 3 -> 2 -> 0

Adjacency list of vertex 2:
HEAD(2) -> 3 -> 1

Adjacency list of vertex 3:
HEAD(3) -> 4 -> 2 -> 1

Adjacency list of vertex 4:
HEAD(4) -> 3 -> 1 -> 0

*/
